app/src/main/java/com/randomappsinc/studentpicker/common/VoiceRecognizerDialogFragment.java

```diff
@@ -0,0 +1,161 @@
+    @BindView(R.id.try_again) TextView tryAgain;
+
+    private SpeechRecognizer mSpeechRecognizer;
+    private Intent mSpeechRecognizerIntent;
+    private Context context;
+    private VoiceRecognizerInterface signal;
+
+    public VoiceRecognizerDialogFragment(Context context, VoiceRecognizerInterface signal) {
+        this.context = context;
+        this.signal = signal;
+    }
+
+    public VoiceRecognizerDialogFragment() {
+    }
+
+    @Override
+    public void onResume() {
+        super.onResume();
+        startListening();
+        getDialog().getWindow().setLayout(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.WRAP_CONTENT);
+    }
+
+    @Nullable
+    @Override
+    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
+        View view = inflater.inflate(R.layout.voice_recognizble_layout, container, false);
+        ButterKnife.bind(this, view);
+        getDialog().getWindow().getAttributes().windowAnimations = R.style.SpeechDialogAnimation;
+
+        mSpeechRecognizer = SpeechRecognizer.createSpeechRecognizer(context);
+        mSpeechRecognizerIntent = new Intent(RecognizerIntent.ACTION_RECOGNIZE_SPEECH);
+        mSpeechRecognizerIntent.putExtra(RecognizerIntent.EXTRA_LANGUAGE_MODEL, RecognizerIntent.LANGUAGE_MODEL_WEB_SEARCH);
+        mSpeechRecognizerIntent.putExtra(RecognizerIntent.EXTRA_LANGUAGE, "en");
+        mSpeechRecognizerIntent.putExtra(RecognizerIntent.EXTRA_PARTIAL_RESULTS, true);
+        mSpeechRecognizerIntent.putExtra(RecognizerIntent.EXTRA_CALLING_PACKAGE, context.getPackageName());
+        mSpeechRecognizer.setRecognitionListener(this);
+        mSpeechRecognizer.startListening(mSpeechRecognizerIntent);
+        return view;
+    }
+
+    private void startListening() {
+        mSpeechRecognizer.setRecognitionListener(this);
+        mSpeechRecognizer.startListening(mSpeechRecognizerIntent);
+        changeUIStateToListening();
+    }
+
+    @Override
+    public void onDestroy() {
+        super.onDestroy();
+        if (mSpeechRecognizer != null) {
+            mSpeechRecognizer.destroy();
+        }
+    }
+
+    @Override
+    public void onBeginningOfSpeech() {
+        message.setText(getString(R.string.waiting));
+    }
+
+    @Override
+    public void onBufferReceived(byte[] buffer) {
+    }
+
+    @Override
+    public void onEndOfSpeech() {
+    }
+
+    @Override
+    public void onError(int error) {
+        if (error == 7) {
+            changeUIStateToRetry();
+        }
+    }
+
+    @Override
+    public void onEvent(int eventType, Bundle params) {
+    }
+
+    @Override
+    public void onPartialResults(Bundle partialResults) {
+    }
+
+    @Override
+    public void onReadyForSpeech(Bundle params) {
+    }
+
+    @Override
+    public void onResults(Bundle results) {
+        ArrayList<String> matches = results.getStringArrayList(SpeechRecognizer.RESULTS_RECOGNITION);
```

Comment:
nit: We should always code to an interface when we can. A general principle is to "hide" underlying implementation as much as possible.
```suggestion
        List<String> matches = results.getStringArrayList(SpeechRecognizer.RESULTS_RECOGNITION);
```